dso-l1/src/main/java/com/tc/object/ClientEntityManagerImpl.java

```diff
@@ -120,6 +120,19 @@ public boolean checkBusy() {
     }
   } 
   
+  private synchronized boolean enqueueMessage(InFlightMessage msg) {
+    if (!this.stateManager.isRunning()) {
+      return false;
+    }
+    if (!requestTickets.messagePendingSlotAvailable()) {
+      throw new IllegalStateException("Output queue is full");
```

Comment:
This seems risky to me, there is no better way to handle it?

Reply:
`IllegalStateException` will eventually be caught in `EntityClientEndpointImpl.invoke()` and call `InvocationCallback.failure()`.

This seemed like the only thing to do in async mode: report to the callback that the output queue is full. Do you have a better proposition in mind?

Reply:
I think it might be better to throw this directly at the caller rather than report it. It's really a distinct *kind* of failure, right?

Reply:
Is it really that distinct? After all, you pushed a message through the API and it failed to be processed. How is that any different than if the client could failed to deliver the message to the server?

Reply:
I guess my main justification here is that it's synchronously generated which might have implications for the callback logic (although that would be grounds for throwing all synchronously generated exceptions).

Reply:
Would you prefer an API where you have to handle exceptions both with a try-catch around `AsyncInvocationBuilder.invoke()` _and_ in `InvocationCallback.failure()` or just in the latter? My personal preference goes for the 2nd option.

If we can't agree, we should involve more team members.

Reply:
Maybe throw RejectedExecutionException which gets thrown back to the caller of ayncInvoke()?  Seems better to let the caller know that nothing ever got started rather than having the callback handle it.  My concern is that the linear sequence gets all messed up if you don't handle this at the caller of asyncInvoke.  It might not matter in most cases but I could see situations where it matters.

Reply:
Okay, I trust your instinct. I'll make the code throw `RejectedExecutionException` and declare that as `throws` all along the stack.